In [ ]:
import numpy as np
import pandas as pd

In [ ]:
datacovid = pd.read_excel('/content/DataCovid19Excel.xlsx')

In [ ]:
datacovid

Date Location ISO Code           Location  New Cases  \
0     2020-01-03 00:00:00             ID-JK        DKI Jakarta          3   
1     2020-01-03 00:00:00             ID-JB         Jawa Barat          3   
2     2020-02-03 00:00:00             ID-JK        DKI Jakarta          2   
3     2020-02-03 00:00:00               IDN          Indonesia          2   
4     2020-02-03 00:00:00             ID-JB         Jawa Barat          0   
...                   ...               ...                ...        ...   
3488            6/30/2020             ID-SG  Sulawesi Tenggara         18   
3489            6/30/2020             ID-SA     Sulawesi Utara         27   
3490            6/30/2020             ID-SB     Sumatera Barat          1   
3491            6/30/2020             ID-SS   Sumatera Selatan         26   
3492            6/30/2020             ID-SU     Sumatera Utara         71   

      New Deaths  New Recovered  New Active Cases  Total Cases  Total Deaths  \
0              0              0                 3            3             0   
1              0              0                 3            3             0   
2              0              0                 2            5             0   
3              0              0                 2            2             0   
4              0              0                 0            3             0   
...          ...            ...               ...          ...           ...   
3488           0             12                 6          363             6   
3489           1             10                16         1121            81   
3490           0             12               -11          726            31   
3491           7              5                14         2048            88   
3492           1              5                65         1544            92   

      Total Recovered  ...   Longitude      Latitude  New Cases per Million  \
0                   0  ...  1068361183 -6.204699e+09                   0.28   
1                   0  ...  1076037083 -6.920432e+09                   0.07   
2                   0  ...  1068361183 -6.204699e+09                   0.18   
3                   0  ...   113921327 -7.892750e-01                   0.01   
4                   0  ...  1076037083 -6.920432e+09                   0.00   
...               ...  ...         ...           ...                    ...   
3488              234  ...   122070311 -4.124689e+09                   6.83   
3489              186  ...  1245212396  1.259638e+09                  10.22   
3490              600  ...  1004650624 -8.502532e-01                   0.18   
3491             1024  ...  1041694647 -3.216212e+09                   3.16   
3492              406  ...  9905196442  2.191894e+09                   4.77   

     Total Cases per Million New Deaths per Million Total Deaths per Million  \
0                       0.28                   0.00                     0.00   
1                       0.07                   0.00                     0.00   
2                       0.46                   0.00                     0.00   
3                       0.01                   0.00                     0.00   
4                       0.07                   0.00                     0.00   
...                      ...                    ...                      ...   
3488                  137.74                   0.00                     2.28   
3489                  424.32                   0.38                    30.66   
3490                  131.54                   0.00                     5.62   
3491                  249.22                   0.85                    10.71   
3492                  103.80                   0.07                     6.18   

     Case Fatality Rate Case Recovered Rate Growth Factor of New Cases  \
0                 0.00%               0.00%                        NaN   
1                 0.00%               0.00%                        NaN   
2  

In [ ]:
datacovid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3493 entries, 0 to 3492
Data columns (total 37 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Date                         3493 non-null   object 
 1   Location ISO Code            3493 non-null   object 
 2   Location                     3493 non-null   object 
 3   New Cases                    3493 non-null   int64  
 4   New Deaths                   3493 non-null   int64  
 5   New Recovered                3493 non-null   int64  
 6   New Active Cases             3493 non-null   int64  
 7   Total Cases                  3493 non-null   int64  
 8   Total Deaths                 3493 non-null   int64  
 9   Total Recovered              3493 non-null   int64  
 10  Total Active Cases           3493 non-null   int64  
 11  Location Level               3493 non-null   object 
 12  City or Regency              0 non-null      float64
 13  Province          

In [ ]:
datacovid.columns

Index(['Date', 'Location ISO Code', 'Location', 'New Cases', 'New Deaths',
       'New Recovered', 'New Active Cases', 'Total Cases', 'Total Deaths',
       'Total Recovered', 'Total Active Cases', 'Location Level',
       'City or Regency', 'Province', 'Country', 'Continent', 'Island',
       'Time Zone', 'Special Status', 'Total Regencies', 'Total Cities',
       'Total Districts', 'Total Urban Villages', 'Total Rural Villages',
       'Area (km2)', 'Population', 'Population Density', 'Longitude',
       'Latitude', 'New Cases per Million', 'Total Cases per Million',
       'New Deaths per Million', 'Total Deaths per Million',
       'Case Fatality Rate', 'Case Recovered Rate',
       'Growth Factor of New Cases', 'Growth Factor of New Deaths'],
      dtype='object')

In [ ]:
import re

def to_lower_camel_case(col):
    # Hilangkan spasi di depan/belakang
    col = col.strip()
    # Pecah berdasarkan spasi atau underscore
    parts = re.split(r'[ _]+', col)
    # Part pertama lowercase, sisanya Capitalize
    return parts[0].lower() + ''.join(word.capitalize() for word in parts[1:])

# Terapkan ke seluruh kolom dataframe
datacovid.columns = [to_lower_camel_case(col) for col in datacovid.columns]

In [ ]:
datacovid.columns

Index(['date', 'locationIsoCode', 'location', 'newCases', 'newDeaths',
       'newRecovered', 'newActiveCases', 'totalCases', 'totalDeaths',
       'totalRecovered', 'totalActiveCases', 'locationLevel', 'cityOrRegency',
       'province', 'country', 'continent', 'island', 'timeZone',
       'specialStatus', 'totalRegencies', 'totalCities', 'totalDistricts',
       'totalUrbanVillages', 'totalRuralVillages', 'area(km2)', 'population',
       'populationDensity', 'longitude', 'latitude', 'newCasesPerMillion',
       'totalCasesPerMillion', 'newDeathsPerMillion', 'totalDeathsPerMillion',
       'caseFatalityRate', 'caseRecoveredRate', 'growthFactorOfNewCases',
       'growthFactorOfNewDeaths'],
      dtype='object')

In [ ]:
datacovid = datacovid.drop(['cityOrRegency'], axis=1)

In [ ]:
datacovid['province'].unique()

array(['DKI Jakarta', 'Jawa Barat', nan, 'Banten', 'Sulawesi Tenggara',
       'Kalimantan Timur', 'Daerah Istimewa Yogyakarta', 'Sumatera Utara',
       'Jawa Tengah', 'Jawa Timur', 'Sulawesi Selatan', 'Maluku', 'Papua',
       'Maluku Utara', 'Aceh', 'Kalimantan Tengah', 'Lampung',
       'Sulawesi Tengah', 'Sumatera Barat', 'Papua Barat',
       'Kalimantan Utara', 'Sulawesi Barat', 'Sumatera Selatan', 'Bali',
       'Kalimantan Barat', 'Riau', 'Jambi', 'Kalimantan Selatan',
       'Kepulauan Bangka Belitung', 'Kepulauan Riau', 'Bengkulu',
       'Nusa Tenggara Barat', 'Sulawesi Utara', 'Nusa Tenggara Timur',
       'Gorontalo'], dtype=object)

In [ ]:
dataclust = datacovid.groupby('province')[['totalCases','totalDeaths','totalRecovered']].max().reset_index()

In [ ]:
dataclust

province  totalCases  totalDeaths  totalRecovered
0                         Aceh          80            3              26
1                         Bali        1484           12             791
2                       Banten        1388           78             639
3                     Bengkulu         125           12              89
4                  DKI Jakarta       10940          612            6473
5   Daerah Istimewa Yogyakarta         298            8             263
6                    Gorontalo         249           10             210
7                        Jambi         116            0              69
8                   Jawa Barat        3215          171            1603
9                  Jawa Tengah        3822          142            1152
10                  Jawa Timur       12087          892            4000
11            Kalimantan Barat         318            2             267
12          Kalimantan Selatan        3147          189             766
13           Kalimantan Tengah         890           54             346
14            Kalimantan Timur         515            7             383
15            Kalimantan Utara         206            2             157
16   Kepulauan Bangka Belitung         152            2             133
17              Kepulauan Riau         288           15             248
18                     Lampung         189           12             143
19                      Maluku         742           16             263
20                Maluku Utara         718           31             105
21         Nusa Tenggara Barat        1232           60             816
22         Nusa Tenggara Timur         113            1              40
23                       Papua        1749           15             856
24                 Papua Barat         239            3             153
25                        Riau         225           10             176
26              Sulawesi Barat         114            2              84
27            Sulawesi Selatan        5083          167            1838
28             Sulawesi Tengah         189            5             153
29           Sulawesi Tenggara         363            6             234
30              Sulawesi Utara        1121           81             186
31              Sumatera Barat         726           31             600
32            Sumatera Selatan        2048           88            1024
33              Sumatera Utara        1544           92             406

**Mengelompokkan wilayah berdasarkan tingkat keparahan COVID-19:**
| Cluster | Tingkat Keparahan       |
| ------- | ----------------------- |
| 1       | Rendah                  |
| 2       | Menengah-rendah         |
| 3       | Menengah-tinggi         |
| 4       | Sangat tinggi           |

---

# 🟦 **Cluster 1 — Wilayah dengan Kasus Rendah**

**Karakteristik utama:**

* Total kasus sangat kecil (±298)
* Kematian rendah (±9)
* Recovered cukup rendah (±190)

📌 **Interpretasi:**
Cluster ini berisi **wilayah dengan penyebaran COVID-19 sangat rendah**.
Kemungkinan wilayah:

* Wilayah berpenduduk kecil
* Wilayah terpencil
* Mobilitas rendah
* Kasus yang terdeteksi sedikit

Ini merupakan **kategori risiko rendah**.

---

# 🟩 **Cluster 2 — Wilayah dengan Kasus Menengah-Rendah**

**Ciri-ciri:**

* Kasus sedang (±1432)
* Kematian relatif kecil (±60)
* Tingkat pemulihan menengah (±633)

📌 **Interpretasi:**
Wilayah dalam cluster ini mengalami **penyebaran lebih besar daripada cluster 0**, tetapi belum parah.

Biasanya wilayah seperti:

* Kota/kabupaten dengan populasi besar tapi cukup terkendali
* Punya fasilitas kesehatan cukup baik
* Ada upaya pengendalian yang efektiv

Cluster ini bisa disebut **risiko menengah**.

---

# 🟧 **Cluster 3 — Wilayah dengan Kasus Menengah-Tinggi**

**Ciri:**

* Total kasus tinggi (±3816)
* Kematian sedang (±167)
* Recovered relatif tinggi (±1339)

📌 **Interpretasi:**
Cluster ini menggambarkan wilayah **dengan beban kasus besar**, tetapi:

* Fatalitas tidak setinggi cluster 1
* Recovered cukup besar → penanganan cukup baik

Cluster ini mencerminkan wilayah:

* Padat penduduk
* Pusat mobilitas tinggi
* Tingkat pelacakan/testing lebih baik

Termasuk kategori **risiko tinggi**, namun masih terkendali dibanding cluster 1.

---

# 🟥 **Cluster 4 — Wilayah dengan Kasus Sangat Tinggi (Hotspot)**

**Ciri:**

* Kasus **sangat besar** (±11.5 ribu)
* Kematian tinggi (±752)
* Recovered juga tinggi (±5236)

📌 **Interpretasi:**
Cluster ini berisi **wilayah dengan wabah paling parah**, biasanya:

* Provinsi/kota besar (misalnya ibu kota, daerah metropolitan)
* Mobilitas penduduk sangat tinggi
* Kepadatan tinggi
* Testing tinggi → banyak kasus terlaporkan
* Fasilitas kesehatan kewalahan (ditunjukkan oleh tingginya kematian)

Kita bisa menyebut cluster ini sebagai **wilayah kritis / epicenter**.



In [ ]:
X = dataclust[['totalCases','totalDeaths','totalRecovered']]

In [ ]:
from sklearn.cluster import KMeans

In [ ]:
kmeans = KMeans(n_clusters = 4, random_state = 0, init = 'k-means++', n_init='auto', max_iter = 500)
kmeans.fit(X)
clust_kmeans = kmeans.fit_predict(X)

In [ ]:
dataclust['cluster'] = clust_kmeans
dataclust

province  totalCases  totalDeaths  totalRecovered  \
0                         Aceh          80            3              26   
1                         Bali        1484           12             791   
2                       Banten        1388           78             639   
3                     Bengkulu         125           12              89   
4                  DKI Jakarta       10940          612            6473   
5   Daerah Istimewa Yogyakarta         298            8             263   
6                    Gorontalo         249           10             210   
7                        Jambi         116            0              69   
8                   Jawa Barat        3215          171            1603   
9                  Jawa Tengah        3822          142            1152   
10                  Jawa Timur       12087          892            4000   
11            Kalimantan Barat         318            2             267   
12          Kalimantan Selatan        3147          189             766   
13           Kalimantan Tengah         890           54             346   
14            Kalimantan Timur         515            7             383   
15            Kalimantan Utara         206            2             157   
16   Kepulauan Bangka Belitung         152            2             133   
17              Kepulauan Riau         288           15             248   
18                     Lampung         189           12             143   
19                      Maluku         742           16             263   
20                Maluku Utara         718           31             105   
21         Nusa Tenggara Barat        1232           60             816   
22         Nusa Tenggara Timur         113            1              40   
23                       Papua        1749           15             856   
24                 Papua Barat         239            3             153   
25                        Riau         225           10             176   
26              Sulawesi Barat         114            2              84   
27            Sulawesi Selatan        5083          167            1838   
28             Sulawesi Tengah         189            5             153   
29           Sulawesi Tenggara         363            6             234   
30              Sulawesi Utara        1121           81             186   
31              Sumatera Barat         726           31             600   
32            Sumatera Selatan        2048           88            1024   
33              Sumatera Utara        1544           92             406   

    cluster  
0         0  
1         3  
2         3  
3         0  
4         1  
5         0  
6         0  
7         0  
8         2  
9         2  
10        1  
11        0  
12        2  
13        3  
14        0  
15        0  
16        0  
17        0  
18        0  
19        0  
20        0  
21        3  
22        0  
23        3  
24        0  
25        0  
26        0  
27        2  
28        0  
29        0  
30        3  
31        0  
32        3  
33        3

In [ ]:
centroid_kmeans = dataclust.groupby('cluster', as_index=False)[['totalCases','totalDeaths','totalRecovered']].mean()
centroid_kmeans

cluster  totalCases  totalDeaths  totalRecovered
0        0      298.25         8.90          189.80
1        1    11513.50       752.00         5236.50
2        2     3816.75       167.25         1339.75
3        3     1432.00        60.00          633.00

In [ ]:
def f(row):
    if row['cluster'] == 0:
        val = 1
    elif row['cluster'] == 3:
        val = 2
    elif row['cluster'] == 2:
        val = 3
    else:
        val = 4
    return val

In [ ]:
dataclust['cluster'] = dataclust.apply(f, axis=1)
dataclust

province  totalCases  totalDeaths  totalRecovered  \
0                         Aceh          80            3              26   
1                         Bali        1484           12             791   
2                       Banten        1388           78             639   
3                     Bengkulu         125           12              89   
4                  DKI Jakarta       10940          612            6473   
5   Daerah Istimewa Yogyakarta         298            8             263   
6                    Gorontalo         249           10             210   
7                        Jambi         116            0              69   
8                   Jawa Barat        3215          171            1603   
9                  Jawa Tengah        3822          142            1152   
10                  Jawa Timur       12087          892            4000   
11            Kalimantan Barat         318            2             267   
12          Kalimantan Selatan        3147          189             766   
13           Kalimantan Tengah         890           54             346   
14            Kalimantan Timur         515            7             383   
15            Kalimantan Utara         206            2             157   
16   Kepulauan Bangka Belitung         152            2             133   
17              Kepulauan Riau         288           15             248   
18                     Lampung         189           12             143   
19                      Maluku         742           16             263   
20                Maluku Utara         718           31             105   
21         Nusa Tenggara Barat        1232           60             816   
22         Nusa Tenggara Timur         113            1              40   
23                       Papua        1749           15             856   
24                 Papua Barat         239            3             153   
25                        Riau         225           10             176   
26              Sulawesi Barat         114            2              84   
27            Sulawesi Selatan        5083          167            1838   
28             Sulawesi Tengah         189            5             153   
29           Sulawesi Tenggara         363            6             234   
30              Sulawesi Utara        1121           81             186   
31              Sumatera Barat         726           31             600   
32            Sumatera Selatan        2048           88            1024   
33              Sumatera Utara        1544           92             406   

    cluster  
0         1  
1         2  
2         2  
3         1  
4         4  
5         1  
6         1  
7         1  
8         3  
9         3  
10        4  
11        1  
12        3  
13        2  
14        1  
15        1  
16        1  
17        1  
18        1  
19        1  
20        1  
21        2  
22        1  
23        2  
24        1  
25        1  
26        1  
27        3  
28        1  
29        1  
30        2  
31        1  
32        2  
33        2

In [ ]:
dataclust

province  totalCases  totalDeaths  totalRecovered  \
0                         Aceh          80            3              26   
1                         Bali        1484           12             791   
2                       Banten        1388           78             639   
3                     Bengkulu         125           12              89   
4                  DKI Jakarta       10940          612            6473   
5   Daerah Istimewa Yogyakarta         298            8             263   
6                    Gorontalo         249           10             210   
7                        Jambi         116            0              69   
8                   Jawa Barat        3215          171            1603   
9                  Jawa Tengah        3822          142            1152   
10                  Jawa Timur       12087          892            4000   
11            Kalimantan Barat         318            2             267   
12          Kalimantan Selatan        3147          189             766   
13           Kalimantan Tengah         890           54             346   
14            Kalimantan Timur         515            7             383   
15            Kalimantan Utara         206            2             157   
16   Kepulauan Bangka Belitung         152            2             133   
17              Kepulauan Riau         288           15             248   
18                     Lampung         189           12             143   
19                      Maluku         742           16             263   
20                Maluku Utara         718           31             105   
21         Nusa Tenggara Barat        1232           60             816   
22         Nusa Tenggara Timur         113            1              40   
23                       Papua        1749           15             856   
24                 Papua Barat         239            3             153   
25                        Riau         225           10             176   
26              Sulawesi Barat         114            2              84   
27            Sulawesi Selatan        5083          167            1838   
28             Sulawesi Tengah         189            5             153   
29           Sulawesi Tenggara         363            6             234   
30              Sulawesi Utara        1121           81             186   
31              Sumatera Barat         726           31             600   
32            Sumatera Selatan        2048           88            1024   
33              Sumatera Utara        1544           92             406   

    cluster  
0         1  
1         2  
2         2  
3         1  
4         4  
5         1  
6         1  
7         1  
8         3  
9         3  
10        4  
11        1  
12        3  
13        2  
14        1  
15        1  
16        1  
17        1  
18        1  
19        1  
20        1  
21        2  
22        1  
23        2  
24        1  
25        1  
26        1  
27        3  
28        1  
29        1  
30        2  
31        1  
32        2  
33        2

In [ ]:
datacovid.to_csv('/content/covid19_recap.csv')

In [ ]:
dataclust.to_csv('/content/covid19_cluster.csv')